### PR TITLE
Support data encoding with IO stream

### LINKED ISSUE #18: Unify data encoding/decoding method
Use data encoding/decoding with IO streams. i.e. encode with binary.Write, decode with binary.Read

### PR DESCRIPTION
**Pull request by zyiou**
fixes #18

### CODE REVIEW

**Comment by srikartati** [comment]
Can you rebase this? It will be easy to review.

File: pkg/entities/record.go [hunk 1]
```diff
@@ -101,91 +102,41 @@ func (d *dataRecord) PrepareRecord() (uint16, error) {
...
-		}
-		binary.BigEndian.PutUint32(bytesToAppend, uint32(v))
-	case Signed64:
-		v, ok := val.(int64)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type int64")
-		}
-		binary.BigEndian.PutUint64(bytesToAppend, uint64(v))
+	case Unsigned8,  Unsigned16, Unsigned32, Unsigned64, Signed8, Signed16, Signed32, Signed64:
+		err = util.Encode(&d.buff, val)
```
Comment (srikartati) [inline pkg/entities/record.go:109]: how are we checking if the val input argument is same type as we expected? util.Encode seems to take in any argument.

File: pkg/entities/record.go [hunk 2]
```diff
@@ -209,75 +158,60 @@ func (d *dataRecord) AddInfoElement(element *InfoElement, val interface{}) (uint
...
 		}
 	default:
 		return 0, fmt.Errorf("API supports only valid information elements with datatypes given in RFC7011")
 	}
 
-	bytesWritten, err := d.buff.Write(bytesToAppend)
+	//bytesWritten, err := d.buff.Write(bytesToAppend)
 	if err != nil {
 		return 0, err
 	}
```
Comment (srikartati) [inline pkg/entities/record.go:183]: were you planning to delete this?

File: pkg/entities/record.go [hunk 3]
```diff
@@ -101,91 +102,41 @@ func (d *dataRecord) PrepareRecord() (uint16, error) {
 
 func (d *dataRecord) AddInfoElement(element *InfoElement, val interface{}) (uint16, error) {
 	d.fieldCount++
-	var bytesToAppend []byte
-	if element.Len != VariableLength {
-		bytesToAppend = make([]byte, element.Len)
-	} else {
-		bytesToAppend = make([]byte, 0)
-	}
+	length := d.buff.Len()
```
Comment (srikartati) [inline pkg/entities/record.go:105]: length -> initialLength?

File: pkg/entities/record.go [hunk 4]
```diff
@@ -209,75 +158,60 @@ func (d *dataRecord) AddInfoElement(element *InfoElement, val interface{}) (uint
...
 }
 
 func (t *templateRecord) PrepareRecord() (uint16, error) {
 	// Add Template Record Header
-	header := make([]byte, 4)
-	binary.BigEndian.PutUint16(header[0:2], t.templateID)
-	binary.BigEndian.PutUint16(header[2:4], t.fieldCount)
-
-	_, err := t.buff.Write(header)
+	length := t.buff.Len()
```
Comment (srikartati) [inline pkg/entities/record.go:189]: same here

File: pkg/entities/record.go [hunk 5]
```diff
@@ -209,75 +158,60 @@ func (d *dataRecord) AddInfoElement(element *InfoElement, val interface{}) (uint
...
+	return uint16(t.buff.Len()-length), nil
 }
 
 func (t *templateRecord) AddInfoElement(element *InfoElement, val interface{}) (uint16, error) {
 	// val could be used to specify smaller length than default? For now assert it to be nil
 	if val != nil {
 		return 0, fmt.Errorf("AddInfoElement(templateRecord) cannot take value %v (nil is expected)", val)
 	}
+	length := t.buff.Len()
 	// Add field specifier
```
Comment (srikartati) [inline pkg/entities/record.go:203]: Please copy the format that was there in decoding here.

File: pkg/entities/record.go [hunk 6]
```diff
@@ -209,75 +158,60 @@ func (d *dataRecord) AddInfoElement(element *InfoElement, val interface{}) (uint
...
-	return uint16(len(header)), nil
+	return uint16(t.buff.Len()-length), nil
 }
 
 func (t *templateRecord) AddInfoElement(element *InfoElement, val interface{}) (uint16, error) {
 	// val could be used to specify smaller length than default? For now assert it to be nil
 	if val != nil {
 		return 0, fmt.Errorf("AddInfoElement(templateRecord) cannot take value %v (nil is expected)", val)
 	}
+	length := t.buff.Len()
```
Comment (srikartati) [inline pkg/entities/record.go:202]: same here.. initialLength

File: pkg/entities/record.go [hunk 7]
```diff
@@ -295,4 +229,4 @@ func (t *templateRecord) GetTemplateElements() []*InfoElement {
 
 func (t *templateRecord) GetMinDataRecordLen() uint16 {
 	return t.minDataRecLength
-}
+}
```
Comment (srikartati) [inline pkg/entities/record.go:232]: new line.

File: pkg/exporter/process_test.go [hunk 8]
```diff
@@ -164,11 +186,11 @@ func TestExportingProcess_SendingTemplateRecordToLocalUDPServer(t *testing.T) {
 
 	// Expect to receive two template headers one from AddRecordAndSendMsg and other from tempRefresh go routine
 	bytesAtServer := <-buffCh
-	assert.Equal(t, len(bytesAtServer), 64)
-	assert.Equal(t, bytesAtServer[20:32], bytesAtServer[52:], "both template messages should be same")
-	firstTemplateBytes := bytesAtServer[:32]
+	assert.Equal(t, 80, len(bytesAtServer))
+	assert.Equal(t, bytesAtServer[20:40], bytesAtServer[60:], "both template messages should be same")
+	firstTemplateBytes := bytesAtServer[:40]
 	// 32 is the size of the IPFIX message including all headers
```
Comment (srikartati) [inline pkg/exporter/process_test.go:192]: change the message size in the comment

File: pkg/exporter/process_test.go [hunk 9]
```diff
@@ -63,10 +67,12 @@ func TestExportingProcess_SendingTemplateRecordToLocalTCPServer(t *testing.T) {
 	// Add template to exporting process
 	reg := registry.NewIanaRegistry()
 	reg.LoadRegistry()
+	antreaReg := registry.NewAntreaRegistry()
+	antreaReg.LoadRegistry()
```
Comment (srikartati) [inline pkg/exporter/process_test.go:71]: You seem to be using the global registry. Do you still need this?

File: pkg/exporter/process_test.go [hunk 10]
```diff
@@ -137,10 +150,12 @@ func TestExportingProcess_SendingTemplateRecordToLocalUDPServer(t *testing.T) {
...
```
Comment (srikartati) [inline pkg/exporter/process_test.go:154]: same here w.r.t global registry.

File: pkg/exporter/process_test.go [hunk 11]
```diff
@@ -216,6 +238,8 @@ func TestExportingProcess_SendingDataRecordToLocalTCPServer(t *testing.T) {
 	// Add data to exporting process
 	reg := registry.NewIanaRegistry()
 	reg.LoadRegistry()
+	antreaReg := registry.NewAntreaRegistry()
+	antreaReg.LoadRegistry()
```
Comment (srikartati) [inline pkg/exporter/process_test.go:242]: same here

File: pkg/util/util.go [hunk 12]
```diff
@@ -20,6 +20,15 @@ import (
 	"io"
 )
 
+// Encode writes data of specified interfaces to io writer
+func Encode(buff io.Writer, inputs ...interface{}) error {
+	var err error
+	for _, in := range inputs {
+		err = binary.Write(buff, binary.BigEndian, in)
```
Comment (srikartati) [inline pkg/util/util.go:27]: have endianness as paramter in both encode and decode functions to increase their utility

File: pkg/util/util.go [hunk 13]
```diff
@@ -20,6 +20,15 @@ import (
 	"io"
 )
 
+// Encode writes data of specified interfaces to io writer
```
Comment (srikartati) [inline pkg/util/util.go:23]: suggestion: Encode writes the multiple inputs of different possible datatypes to the io writer.

File: pkg/exporter/process_test.go (hunk 9 above)
Comment (zyiou) [reply pkg/exporter/process_test.go:71]: Will make the change in global registry PR

File: pkg/util/util.go (hunk 12 above)
Comment (zyiou) [reply pkg/util/util.go:27]: Makes sense. Thanks

File: pkg/entities/record.go (hunk 2 above)
Comment (zyiou) [reply pkg/entities/record.go:183]: It is for handling all possible error from encoding. not planning to delete it.

File: pkg/entities/record.go (hunk 1 above)
Comment (zyiou) [reply pkg/entities/record.go:109]: I will need to handle each case separately then. Thanks for pointing it out.

File: pkg/util/util.go [hunk 14]
```diff
@@ -20,13 +20,16 @@ import (
 	"io"
 )
 
-// Encode writes data of specified interfaces to io writer
-func Encode(buff io.Writer, inputs ...interface{}) error {
+// Encode writes the multiple inputs of different possible datatypes to the io writer.
+func Encode(buff io.Writer, order binary.ByteOrder, inputs ...interface{}) error {
```
Comment (srikartati) [inline pkg/util/util.go:24]: byteOrder is more clear for variable name. Same below.

**Review by srikartati — COMMENTED** [review]
Sorry for late turn around @zyiou .
Main comment: is it possible to merge Addinfoelement of dataset and datarecord? Idea is to use same datarecord object for both encoding and decoding? Utility function can be passed as an argument to the method. Ideally data records should be part of data set. Do you think we can built abstraction like above that is common for encoding and decoding?

**Comment by zyiou** [comment]
Agreed @srikartati . It would be useful for reusing of the code to encoding/decoding and also support multiple records per set in the future. Will do that. Thanks!

File: pkg/entities/record.go [hunk 15]
```diff
@@ -209,75 +205,60 @@ func (d *dataRecord) AddInfoElement(element *InfoElement, val interface{}) (uint
...
-
-			bytesToAppend = append(bytesToAppend, []byte(v)...)
+			err = util.Encode(&d.buff, binary.BigEndian, byte(255), uint16(len(v)), []byte(v))
 		}
 	default:
 		return 0, fmt.Errorf("API supports only valid information elements with datatypes given in RFC7011")
 	}
 
-	bytesWritten, err := d.buff.Write(bytesToAppend)
+	//bytesWritten, err := d.buff.Write(bytesToAppend)
```
Comment (srikartati) [inline pkg/entities/record.go:226]: commented?

**Review by srikartati — COMMENTED** [review]
Hi Yiou, Are you planning to merge AddInfoElements of dataRecord and dataSet in this PR or a separate PR?

File: pkg/entities/record.go (hunk 15 above)
Comment (zyiou) [reply pkg/entities/record.go]: This needs to be removed. Thanks!

**Comment by zyiou** [comment]
Hi Srikar, I have created another PR for merging AddInfoElement and support for multiple data records per data set in exporting process for handling it. #49 Thanks!

**Review by srikartati — APPROVED** [review]
LGTM.

**Pull request merged**